src/util/README.md

```diff
@@ -383,3 +383,45 @@ _not_ be extended - but it may be helpful to be aware of where this happens.
     These are in the process of being removed - no new output should go
     via `std::cout` or `std::cerr`, but should instead use the
     \ref messaget and \ref message_handlert infrastructure.
+
+
+\subsection Graph
+### Graph
+
+Implemented in `src/util/graph.h` as `grapht` class. The class natively defines
+only a directional graph. However, an undirected graph can be emulated by
+inserting for each edge (u, v) also (v, u). A multi-graph is not supported
+though, because parallel edges are not allowed between vertices. 
+ 
+#### Data representation
+
+A graph is defined by a template class `grapht<N>`, where the type `N` defines
+a node of the graph. The class `grapht` stores the nodes in a vector. A user of
+the graph is supposed to access the nodes via their indices in the vector. The
+node type `N` must meet these requirements:
+- It must be default constructible; the common way how to insert a node to the
+  graph is to call method `add_node` which pushes a new default-constructed
+  instance of `N` to the end of the vector and returns its index.
+- It must define a type `edget`, representing type of data attached to any edge
```

Comment:
I wonder why we do it this way; I can't see any obvious reason why the edge type necessarily follows from the node type. It seems the type parameter here doesn't specify the node type, but actually _both_ the node type and the edge type? Maybe it would be worth refactoring the graph to work like this `grapht<graph_configt> { using node_t = typename graph_configt::node_t; using edge_t = graph_configt::edge_t; // ... }`? Just a thought

Reply:
Indeed we just define edges and node types at the same time. But of course this can be improved upon, and improvements are always welcome :-)